toolchain/check/convert.cpp

```diff
@@ -767,6 +768,59 @@ static auto CanUseValueOfInitializer(const SemIR::File& sem_ir,
   return InitReprIsCopyOfValueRepr(sem_ir, type_id);
 }
 
+// Determine whether the given set of qualifiers can be added by a conversion
+// of an expression of the given category.
+static auto CanAddQualifiers(SemIR::TypeQualifiers quals,
+                             SemIR::ExprCategory cat) -> bool {
+  if ((quals & SemIR::TypeQualifiers::MaybeUnformed) !=
+          SemIR::TypeQualifiers::None &&
+      !SemIR::IsRefCategory(cat)) {
+    // `MaybeUnformed(T)` may have a different value representation or
+    // initializing representation from `T`, so only allow it to be added for a
+    // reference expression.
+    // TODO: We should allow converting an initializing expression of type `T`
+    // to `MaybeUnformed(T)` by generating an `InPlaceInit` instruction when
+    // needed.
+    // NOLINTNEXTLINE(readability-simplify-boolean-expr)
+    return false;
+  }
+
+  // `const` and `partial` can always be added.
+  return true;
+}
+
+// Determine whether the given set of qualifiers can be removed by a conversion
+// of an expression of the given category.
+static auto CanRemoveQualifiers(SemIR::TypeQualifiers quals,
+                                SemIR::ExprCategory cat, bool allow_unsafe)
+    -> bool {
+  if ((quals & SemIR::TypeQualifiers::Const) != SemIR::TypeQualifiers::None &&
+      !allow_unsafe && SemIR::IsRefCategory(cat)) {
+    // Removing `const` is an unsafe conversion for a reference expression.
+    return false;
+  }
+
+  if ((quals & SemIR::TypeQualifiers::Partial) != SemIR::TypeQualifiers::None &&
+      (!allow_unsafe || cat == SemIR::ExprCategory::Initializing)) {
+    // TODO: Allow removing `partial` for initializing expressions as a safe
+    // conversion, and initialize the vptr as part of the conversion.
+    return false;
+  }
+
+  if ((quals & SemIR::TypeQualifiers::MaybeUnformed) !=
+          SemIR::TypeQualifiers::None &&
+      (!allow_unsafe || !SemIR::IsRefCategory(cat))) {
+    // As an unsafe conversion, `MaybeUnformed` can be removed from a reference
+    // expression.
+    // TODO: We should allow this for any kind of expression, and convert the
+    // result as needed if the representation of `T` differs from that of
+    // `MaybeUnformed(T)`.
+    return false;
+  }
```

Comment:
Where you say "convert the result", that sounds like a TODO for the call site; is that correct? If so, had you considered putting these `if` and TODOs at the call site instead? I assume that'd mean having `CanRemoveQualifiers` only deal with const.

But this doesn't really matter if you're already working on fixing it, or if it's too ugly to you, just noting that it's taking me a little time to understand if you're expecting the TODOs to remain longer term. If you leave the TODOs here, you might consider specifying call site work, such as "and convert" -> "and the call site should convert"